File: packages/esm-format/src/expression.ts
/**
 * Expression structural operations for the ESM format
 *
 * This module provides utilities for analyzing and manipulating mathematical
 * expressions in the ESM format AST.
 */

import type { Expression, ExpressionNode, Model } from './types.js'

/**
 * Type alias for better readability
 */
export type Expr = Expression

/**
 * Extract all variable references from an expression
 * @param expr Expression to analyze
 * @returns Set of variable names referenced in the expression
 */
export function freeVariables(expr: Expr): Set<string> {
  const variables = new Set<string>()

  if (typeof expr === 'string') {
    variables.add(expr)
  } else if (typeof expr === 'number') {
    // Numbers contain no variables
    return variables
  } else if (typeof expr === 'object' && expr.op) {
    // ExpressionNode - recursively analyze arguments
    for (const arg of expr.args) {
      const childVars = freeVariables(arg)
      childVars.forEach(v => variables.add(v))
    }
  }

  return variables
}

/**
 * Extract free parameters from an expression within a model context
 * @param expr Expression to analyze
 * @param model Model context to determine parameter vs state variables
 * @returns Set of parameter names referenced in the expression
 */
export function freeParameters(expr: Expr, model: Model): Set<string> {
  const allVars = freeVariables(expr)
  const parameters = new Set<string>()

  for (const varName of allVars) {
    const variable = model.variables[varName]
    if (variable && variable.type === 'parameter') {
      parameters.add(varName)
    }
  }

  return parameters
}

/**
 * Check if an expression contains a specific variable
 * @param expr Expression to search
 * @param varName Variable name to look for
 * @returns True if the variable appears in the expression
 */
export function contains(expr: Expr, varName: string): boolean {
  if (typeof expr === 'string') {
    return expr === varName
  } else if (typeof expr === 'number') {
    return false
  } else if (typeof expr === 'object' && expr.op) {
    // ExpressionNode - recursively check arguments
    return expr.args.some(arg => contains(arg, varName))
  }

  return false
}

/**
 * Evaluate an expression numerically with variable bindings
 * @param expr Expression to evaluate
 * @param bindings Map of variable names to their numeric values
 * @returns Numeric result
 * @throws Error if variables are unbound or evaluation fails
 */
export function evaluate(expr: Expr, bindings: Map<string, number>): number {
  if (typeof expr === 'number') {
    return expr
  } else if (typeof expr === 'string') {
    if (bindings.has(expr)) {
      return bindings.get(expr)!
    } else {
      throw new Error(`Unbound variable: ${expr}`)
    }
  } else if (typeof expr === 'object' && expr.op) {
    // ExpressionNode - evaluate based on operator
    const args = expr.args.map(arg => evaluate(arg, bindings))

    switch (expr.op) {
      case '+':
        return args.reduce((sum, val) => sum + val, 0)
      case '-':
        if (args.length === 1) return -args[0]
        return args.reduce((diff, val, idx) => idx === 0 ? val : diff - val)
      case '*':
        return args.reduce((prod, val) => prod * val, 1)
      case '/':
        if (args.length !== 2) throw new Error('Division requires exactly 2 arguments')
        if (args[1] === 0) throw new Error('Division by zero')
        return args[0] / args[1]
      case '^':
        if (args.length !== 2) throw new Error('Exponentiation requires exactly 2 arguments')
        return Math.pow(args[0], args[1])
      case 'exp':
        if (args.length !== 1) throw new Error('exp requires exactly 1 argument')
        return Math.exp(args[0])
      case 'log':
        if (args.length !== 1) throw new Error('log requires exactly 1 argument')
        if (args[0] <= 0) throw new Error('log argument must be positive')
        return Math.log(args[0])
      case 'log10':
        if (args.length !== 1) throw new Error('log10 requires exactly 1 argument')
        if (args[0] <= 0) throw new Error('log10 argument must be positive')
        return Math.log10(args[0])
      case 'sqrt':
        if (args.length !== 1) throw new Error('sqrt requires exactly 1 argument')
        if (args[0] < 0) throw new Error('sqrt argument must be non-negative')
        return Math.sqrt(args[0])
      case 'abs':
        if (args.length !== 1) throw new Error('abs requires exactly 1 argument')
        return Math.abs(args[0])
      case 'sin':
        if (args.length !== 1) throw new Error('sin requires exactly 1 argument')
        return Math.sin(args[0])
      case 'cos':
        if (args.length !== 1) throw new Error('cos requires exactly 1 argument')
        return Math.cos(args[0])
      case 'tan':
        if (args.length !== 1) throw new Error('tan requires exactly 1 argument')
        return Math.tan(args[0])
      case 'asin':
        if (args.length !== 1) throw new Error('asin requires exactly 1 argument')
        if (args[0] < -1 || args[0] > 1) throw new Error('asin argument must be in [-1, 1]')
        return Math.asin(args[0])
      case 'acos':
        if (args.length !== 1) throw new Error('acos requires exactly 1 argument')
        if (args[0] < -1 || args[0] > 1) throw new Error('acos argument must be in [-1, 1]')
        return Math.acos(args[0])
      case 'atan':
        if (args.length !== 1) throw new Error('atan requires exactly 1 argument')
        return Math.atan(args[0])
      case 'atan2':
        if (args.length !== 2) throw new Error('atan2 requires exactly 2 arguments')
        return Math.atan2(args[0], args[1])
      case 'min':
        if (args.length === 0) throw new Error('min requires at least 1 argument')
        return Math.min(...args)
      case 'max':
        if (args.length === 0) throw new Error('max requires at least 1 argument')
        return Math.max(...args)
      case 'floor':
        if (args.length !== 1) throw new Error('floor requires exactly 1 argument')
        return Math.floor(args[0])
      case 'ceil':
        if (args.length !== 1) throw new Error('ceil requires exactly 1 argument')
        return Math.ceil(args[0])
      case 'sign':
        if (args.length !== 1) throw new Error('sign requires exactly 1 argument')
        return Math.sign(args[0])
      case '>':
        if (args.length !== 2) throw new Error('> requires exactly 2 arguments')
        return args[0] > args[1] ? 1 : 0
      case '<':
        if (args.length !== 2) throw new Error('< requires exactly 2 arguments')
        return args[0] < args[1] ? 1 : 0
      case '>=':
        if (args.length !== 2) throw new Error('>= requires exactly 2 arguments')
        return args[0] >= args[1] ? 1 : 0
      case '<=':
        if (args.length !== 2) throw new Error('<= requires exactly 2 arguments')
        return args[0] <= args[1] ? 1 : 0
      case '==':
        if (args.length !== 2) throw new Error('== requires exactly 2 arguments')
        return args[0] === args[1] ? 1 : 0
      case '!=':
        if (args.length !== 2) throw new Error('!= requires exactly 2 arguments')
        return args[0] !== args[1] ? 1 : 0
      case 'and':
        return args.every(x => x !== 0) ? 1 : 0
      case 'or':
        return args.some(x => x !== 0) ? 1 : 0
      case 'not':
        if (args.length !== 1) throw new Error('not requires exactly 1 argument')
        return args[0] === 0 ? 1 : 0
      case 'ifelse':
        if (args.length !== 3) throw new Error('ifelse requires exactly 3 arguments')
        return args[0] !== 0 ? args[1] : args[2]
      default:
        throw new Error(`Unsupported operator: ${expr.op}`)
    }
  }

  throw new Error('Invalid expression type')
}

/**
 * Simplify an expression using basic algebraic rules
 * @param expr Expression to simplify
 * @returns Simplified expression
 */
export function simplify(expr: Expr): Expr {
  if (typeof expr === 'number' || typeof expr === 'string') {
    return expr
  }

  if (typeof expr === 'object' && expr.op) {
    // First simplify all arguments recursively
    const simplifiedArgs = expr.args.map(arg => simplify(arg))

    // Apply simplification rules based on operator
    switch (expr.op) {
      case '+':
        // Remove zeros: x + 0 -> x
        const nonZeroTerms = simplifiedArgs.filter(arg => arg !== 0)
        if (nonZeroTerms.length === 0) return 0
        if (nonZeroTerms.length === 1) return nonZeroTerms[0]

        // Separate constants and variables for partial constant folding
        const constants = nonZeroTerms.filter(arg => typeof arg === 'number') as number[]
        const variables = nonZeroTerms.filter(arg => typeof arg !== 'number')

        // If all terms are constants, return the sum
        if (variables.length === 0) {
          return constants.reduce((sum, val) => sum + val, 0)
        }

        // If there are constants to fold, combine them
        if (constants.length > 1) {
          const constantSum = constants.reduce((sum, val) => sum + val, 0)
          if (constantSum === 0) {
            // If constant sum is zero, just return variables
            return variables.length === 1 ? variables[0] : { ...expr, args: variables as [Expression, ...Expression[]] }
          } else {
            // Include the folded constant with variables
            const finalTerms = [...variables, constantSum]
            return { ...expr, args: finalTerms as [Expression, ...Expression[]] }
          }
        }

        return { ...expr, args: nonZeroTerms as [Expression, ...Expression[]] }

      case '*':
        // Zero multiplication: x * 0 -> 0
        if (simplifiedArgs.some(arg => arg === 0)) return 0

        // Remove ones: x * 1 -> x
        const nonOneFactors = simplifiedArgs.filter(arg => arg !== 1)
        if (nonOneFactors.length === 0) return 1
        if (nonOneFactors.length === 1) return nonOneFactors[0]

        // Separate constants and variables for partial constant folding
        const constantFactors = nonOneFactors.filter(arg => typeof arg === 'number') as number[]
        const variableFactors = nonOneFactors.filter(arg => typeof arg !== 'number')

        // If all factors are constants, return the product
        if (variableFactors.length === 0) {
          return constantFactors.reduce((prod, val) => prod * val, 1)
        }

        // If there are constants to fold, combine them
        if (constantFactors.length > 1) {
          const constantProd = constantFactors.reduce((prod, val) => prod * val, 1)
          if (constantProd === 0) {
            return 0
          } else if (constantProd === 1) {
            // If constant product is one, just return variables
            return variableFactors.length === 1 ? variableFactors[0] : { ...expr, args: variableFactors as [Expression, ...Expression[]] }
          } else {
            // Include the folded constant with variables
            const finalFactors = [...variableFactors, constantProd]
            return { ...expr, args: finalFactors as [Expression, ...Expression[]] }
          }
        }

        return { ...expr, args: nonOneFactors as [Expression, ...Expression[]] }

      case '-':
        if (simplifiedArgs.length === 1) {
          // Unary minus: -(-x) -> x would need deeper analysis
          if (typeof simplifiedArgs[0] === 'number') {
            return -simplifiedArgs[0]
          }
        } else if (simplifiedArgs.length === 2) {
          // Binary subtraction: x - 0 -> x
          if (simplifiedArgs[1] === 0) return simplifiedArgs[0]

          // Constant folding
          if (typeof simplifiedArgs[0] === 'number' && typeof simplifiedArgs[1] === 'number') {
            return simplifiedArgs[0] - simplifiedArgs[1]
          }
        }

        return { ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }

      case '/':
        if (simplifiedArgs.length === 2) {
          // x / 1 -> x
          if (simplifiedArgs[1] === 1) return simplifiedArgs[0]

          // 0 / x -> 0 (assuming x != 0)
          if (simplifiedArgs[0] === 0) return 0

          // Constant folding
          if (typeof simplifiedArgs[0] === 'number' && typeof simplifiedArgs[1] === 'number') {
            if (simplifiedArgs[1] === 0) throw new Error('Division by zero')
            return simplifiedArgs[0] / simplifiedArgs[1]
          }
        }

        return { ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }

      case '^':
        if (simplifiedArgs.length === 2) {
          // x^0 -> 1
          if (simplifiedArgs[1] === 0) return 1

          // x^1 -> x
          if (simplifiedArgs[1] === 1) return simplifiedArgs[0]

          // 0^x -> 0 (assuming x > 0)
          if (simplifiedArgs[0] === 0) return 0

          // 1^x -> 1
          if (simplifiedArgs[0] === 1) return 1

          // Constant folding
          if (typeof simplifiedArgs[0] === 'number' && typeof simplifiedArgs[1] === 'number') {
            return Math.pow(simplifiedArgs[0], simplifiedArgs[1])
          }
        }

        return { ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }

      default:
        // For other operators, just apply constant folding if all args are numeric
        if (simplifiedArgs.every(arg => typeof arg === 'number')) {
          try {
            // Create a temporary bindings map for evaluation
            const tempBindings = new Map<string, number>()
            return evaluate({ ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }, tempBindings)
          } catch {
            // If evaluation fails, return the expression with simplified args
            return { ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }
          }
        }

        return { ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }
    }
  }

  return expr
}

File: packages/esm_format/src/esm_format/expression.py
"""
Expression manipulation and analysis functions.
"""

from typing import Dict, Set, Union, Optional, List
from .esm_types import Expr, ExprNode, Model, ReactionSystem, ModelVariable
import sympy as sp


def free_variables(expr: Expr) -> Set[str]:
    """
    Extract all free variables from an expression.

    Args:
        expr: Expression to analyze

    Returns:
        Set of variable names found in the expression
    """
    if isinstance(expr, str):
        # String is a variable name
        return {expr}
    elif isinstance(expr, (int, float)):
        # Numbers have no variables
        return set()
    elif isinstance(expr, ExprNode):
        # Recursively collect variables from all arguments
        variables = set()
        for arg in expr.args:
            variables.update(free_variables(arg))
        return variables
    else:
        # Unknown type, assume no variables
        return set()


def contains(expr: Expr, var_name: str) -> bool:
    """
    Check if an expression contains a specific variable.

    Args:
        expr: Expression to search in
        var_name: Variable name to look for

    Returns:
        True if variable is found, False otherwise
    """
    return var_name in free_variables(expr)


def evaluate(expr: Expr, bindings: Dict[str, float]) -> float:
    """
    Evaluate an expression with given variable bindings.

    Args:
        expr: Expression to evaluate
        bindings: Dictionary mapping variable names to values

    Returns:
        Numerical result of evaluation

    Raises:
        ValueError: If unbound variables are encountered
        TypeError: If unsupported operations are encountered
    """
    if isinstance(expr, (int, float)):
        return float(expr)
    elif isinstance(expr, str):
        if expr in bindings:
            return bindings[expr]
        else:
            raise ValueError(f"Unbound variable: {expr}")
    elif isinstance(expr, ExprNode):
        # Evaluate arguments first
        arg_values = [evaluate(arg, bindings) for arg in expr.args]

        # Apply operation
        if expr.op == "+":
            return sum(arg_values)
        elif expr.op == "-":
            if len(arg_values) == 1:
                return -arg_values[0]
            elif len(arg_values) == 2:
                return arg_values[0] - arg_values[1]
            else:
                raise TypeError(f"Invalid number of arguments for subtraction: {len(arg_values)}")
        elif expr.op == "*":
            result = 1.0
            for value in arg_values:
                result *= value
            return result
        elif expr.op == "/":
            if len(arg_values) != 2:
                raise TypeError(f"Division requires exactly 2 arguments, got {len(arg_values)}")
            if arg_values[1] == 0:
                raise ValueError("Division by zero")
            return arg_values[0] / arg_values[1]
        elif expr.op == "^" or expr.op == "**":
            if len(arg_values) != 2:
                raise TypeError(f"Power requires exactly 2 arguments, got {len(arg_values)}")
            return arg_values[0] ** arg_values[1]
        elif expr.op == "log":
            if len(arg_values) != 1:
                raise TypeError(f"Logarithm requires exactly 1 argument, got {len(arg_values)}")
            if arg_values[0] <= 0:
                raise ValueError("Logarithm of non-positive number")
            import math
            return math.log(arg_values[0])
        elif expr.op == "exp":
            if len(arg_values) != 1:
                raise TypeError(f"Exponential requires exactly 1 argument, got {len(arg_values)}")
            import math
            return math.exp(arg_values[0])
        elif expr.op == "sin":
            if len(arg_values) != 1:
                raise TypeError(f"Sine requires exactly 1 argument, got {len(arg_values)}")
            import math
            return math.sin(arg_values[0])
        elif expr.op == "cos":
            if len(arg_values) != 1:
                raise TypeError(f"Cosine requires exactly 1 argument, got {len(arg_values)}")
            import math
            return math.cos(arg_values[0])
        else:
            raise TypeError(f"Unsupported operation: {expr.op}")
    else:
        raise TypeError(f"Unsupported expression type: {type(expr)}")


def simplify(expr: Expr) -> Expr:
    """
    Simplify an expression by performing constant folding and basic algebraic simplifications.

    Args:
        expr: Expression to simplify

    Returns:
        Simplified expression
    """
    if isinstance(expr, (int, float, str)):
        # Atomic expressions don't need simplification
        return expr
    elif isinstance(expr, ExprNode):
        # First, simplify all arguments recursively
        simplified_args = [simplify(arg) for arg in expr.args]

        # Check if all arguments are constants
        all_constants = all(isinstance(arg, (int, float)) for arg in simplified_args)

        if all_constants and len(simplified_args) > 0:
            # If all arguments are constants, evaluate the expression
            try:
                # Create a temporary ExprNode with simplified args for evaluation
                temp_expr = ExprNode(op=expr.op, args=simplified_args)
                return evaluate(temp_expr, {})
            except (ValueError, TypeError, ZeroDivisionError):
                # If evaluation fails, return the expression with simplified args
                return ExprNode(op=expr.op, args=simplified_args, wrt=expr.wrt, dim=expr.dim)

        # Apply specific simplification rules
        if expr.op == "+":
            # Remove zeros and combine constants
            non_zero_args = []
            constant_sum = 0
            has_constants = False

            for arg in simplified_args:
                if isinstance(arg, (int, float)):
                    if arg != 0:
                        constant_sum += arg
                        has_constants = True
                else:
                    non_zero_args.append(arg)

            # Add back the constant sum if non-zero or if there are no other terms
            if has_constants and (constant_sum != 0 or len(non_zero_args) == 0):
                non_zero_args.append(constant_sum)

            if len(non_zero_args) == 0:
                return 0
            elif len(non_zero_args) == 1:
                return non_zero_args[0]
            else:
                return ExprNode(op=expr.op, args=non_zero_args, wrt=expr.wrt, dim=expr.dim)

        elif expr.op == "*":
            # Remove ones, handle zeros, and combine constants
            non_one_args = []
            constant_product = 1
            has_constants = False

            for arg in simplified_args:
                if isinstance(arg, (int, float)):
                    if arg == 0:
                        return 0  # Anything times zero is zero
                    elif arg != 1:
                        constant_product *= arg
                        has_constants = True
                else:
                    non_one_args.append(arg)

            # Add back the constant product if not one or if there are no other terms
            if has_constants and (constant_product != 1 or len(non_one_args) == 0):
                non_one_args.append(constant_product)

            if len(non_one_args) == 0:
                return 1
            elif len(non_one_args) == 1:
                return non_one_args[0]
            else:
                return ExprNode(op=expr.op, args=non_one_args, wrt=expr.wrt, dim=expr.dim)

        elif expr.op == "^" or expr.op == "**":
            # Handle special cases like x^0, x^1, 1^y, 0^y
            if len(simplified_args) == 2:
                base, exponent = simplified_args
                if isinstance(exponent, (int, float)) and exponent == 0:
                    return 1  # x^0 = 1
                elif isinstance(exponent, (int, float)) and exponent == 1:
                    return base  # x^1 = x
                elif isinstance(base, (int, float)) and base == 1:
                    return 1  # 1^y = 1
                elif isinstance(base, (int, float)) and base == 0:
                    return 0  # 0^y = 0 (for positive y)

        # If no specific simplifications apply, return with simplified args
        return ExprNode(op=expr.op, args=simplified_args, wrt=expr.wrt, dim=expr.dim)
    else:
        return expr


def to_sympy(expr: Expr, symbol_map: Optional[Dict[str, sp.Symbol]] = None) -> sp.Expr:
    """
    Convert ESM expression to SymPy expression.

    Args:
        expr: ESM expression to convert
        symbol_map: Optional mapping from variable names to SymPy symbols

    Returns:
        SymPy expression

    Raises:
        TypeError: If unsupported expression type or operation is encountered
    """
    if symbol_map is None:
        symbol_map = {}

    if isinstance(expr, (int, float)):
        return sp.sympify(expr)
    elif isinstance(expr, str):
        # Variable name - create symbol if not in map
        if expr not in symbol_map:
            symbol_map[expr] = sp.Symbol(expr)
        return symbol_map[expr]
    elif isinstance(expr, ExprNode):
        # Convert arguments recursively
        sympy_args = [to_sympy(arg, symbol_map) for arg in expr.args]

        # Map operations to SymPy
        if expr.op == "+":
            return sum(sympy_args) if sympy_args else 0
        elif expr.op == "-":
            if len(sympy_args) == 1:
                return -sympy_args[0]
            elif len(sympy_args) == 2:
                return sympy_args[0] - sympy_args[1]
            else:
                raise TypeError(f"Invalid number of arguments for subtraction: {len(sympy_args)}")
        elif expr.op == "*":
            result = 1
            for arg in sympy_args:
                result *= arg
            return result
        elif expr.op == "/":
            if len(sympy_args) != 2:
                raise TypeError(f"Division requires exactly 2 arguments, got {len(sympy_args)}")
            return sympy_args[0] / sympy_args[1]
        elif expr.op in ("^", "**"):
            if len(sympy_args) != 2:
                raise TypeError(f"Power requires exactly 2 arguments, got {len(sympy_args)}")
            return sympy_args[0] ** sympy_args[1]
        elif expr.op == "exp":
            if len(sympy_args) != 1:
                raise TypeError(f"Exponential requires exactly 1 argument, got {len(sympy_args)}")
            return sp.exp(sympy_args[0])
        elif expr.op == "log":
            if len(sympy_args) != 1:
                raise TypeError(f"Logarithm requires exactly 1 argument, got {len(sympy_args)}")
            return sp.log(sympy_args[0])
        elif expr.op == "sin":
            if len(sympy_args) != 1:
                raise TypeError(f"Sine requires exactly 1 argument, got {len(sympy_args)}")
            return sp.sin(sympy_args[0])
        elif expr.op == "cos":
            if len(sympy_args) != 1:
                raise TypeError(f"Cosine requires exactly 1 argument, got {len(sympy_args)}")
            return sp.cos(sympy_args[0])
        elif expr.op == "D" and expr.wrt:
            # Derivative operation
            if len(sympy_args) != 1:
                raise TypeError(f"Derivative requires exactly 1 argument, got {len(sympy_args)}")
            wrt_symbol = to_sympy(expr.wrt, symbol_map)
            return sp.Derivative(sympy_args[0], wrt_symbol)
        elif expr.op == "Pre":
            # Previous value operator - represent as a function
            if len(sympy_args) != 1:
                raise TypeError(f"Pre operator requires exactly 1 argument, got {len(sympy_args)}")
            return sp.Function('Pre')(sympy_args[0])
        elif expr.op == "ifelse":
            # Conditional operator - map to Piecewise
            if len(sympy_args) != 3:
                raise TypeError(f"ifelse requires exactly 3 arguments (condition, true_val, false_val), got {len(sympy_args)}")
            condition, true_val, false_val = sympy_args
            return sp.Piecewise((true_val, condition), (false_val, True))
        elif expr.op == "grad":
            # Gradient operator - represent as derivative for now
            if len(sympy_args) != 1:
                raise TypeError(f"Gradient requires exactly 1 argument, got {len(sympy_args)}")
            # For gradient, we need the dimension info
            if expr.dim:
                dim_symbol = to_sympy(expr.dim, symbol_map)
                return sp.Derivative(sympy_args[0], dim_symbol)
            else:
                # No dimension specified, just return the expression
                return sympy_args[0]
        else:
            raise TypeError(f"Unsupported operation: {expr.op}")
    else:
        raise TypeError(f"Unsupported expression type: {type(expr)}")


def from_sympy(sympy_expr: sp.Expr) -> Expr:
    """
    Convert SymPy expression back to ESM expression.

    Args:
        sympy_expr: SymPy expression to convert

    Returns:
        ESM expression

    Raises:
        TypeError: If unsupported SymPy expression type is encountered
    """
    if isinstance(sympy_expr, (sp.Integer, sp.Rational, sp.Float)):
        return float(sympy_expr)
    elif isinstance(sympy_expr, sp.Symbol):
        return str(sympy_expr)
    elif isinstance(sympy_expr, sp.Add):
        # Addition
        args = [from_sympy(arg) for arg in sympy_expr.args]
        return ExprNode(op="+", args=args)
    elif isinstance(sympy_expr, sp.Mul):
        # Multiplication
        args = [from_sympy(arg) for arg in sympy_expr.args]
        return ExprNode(op="*", args=args)
    elif isinstance(sympy_expr, sp.Pow):
        # Power
        base, exp = sympy_expr.args
        return ExprNode(op="^", args=[from_sympy(base), from_sympy(exp)])
    elif isinstance(sympy_expr, sp.exp):
        # Exponential
        return ExprNode(op="exp", args=[from_sympy(sympy_expr.args[0])])
    elif isinstance(sympy_expr, sp.log):
        # Logarithm
        return ExprNode(op="log", args=[from_sympy(sympy_expr.args[0])])
    elif isinstance(sympy_expr, sp.sin):
        # Sine
        return ExprNode(op="sin", args=[from_sympy(sympy_expr.args[0])])
    elif isinstance(sympy_expr, sp.cos):
        # Cosine
        return ExprNode(op="cos", args=[from_sympy(sympy_expr.args[0])])
    elif isinstance(sympy_expr, sp.Derivative):
        # Derivative
        expr_arg, wrt_arg = sympy_expr.args[0], sympy_expr.args[1]
        return ExprNode(op="D", args=[from_sympy(expr_arg)], wrt=str(wrt_arg))
    elif isinstance(sympy_expr, sp.Piecewise):
        # Piecewise -> ifelse (assuming 2 pieces)
        pieces = sympy_expr.args
        if len(pieces) == 2:
            (true_val, condition), (false_val, _) = pieces
            return ExprNode(op="ifelse", args=[from_sympy(condition), from_sympy(true_val), from_sympy(false_val)])
        else:
            # For more complex piecewise, just return the first expression for now
            return from_sympy(pieces[0][0])
    elif hasattr(sympy_expr, 'func') and str(sympy_expr.func) == 'Pre':
        # Pre function
        return ExprNode(op="Pre", args=[from_sympy(sympy_expr.args[0])])
    elif isinstance(sympy_expr, sp.Function):
        # Generic function - check if it's one we know
        func_name = str(sympy_expr.func)
        if func_name == 'Pre':
            return ExprNode(op="Pre", args=[from_sympy(arg) for arg in sympy_expr.args])
        else:
            # Unknown function - represent as operation with function name
            return ExprNode(op=func_name, args=[from_sympy(arg) for arg in sympy_expr.args])
    elif sympy_expr.is_number:
        # Numeric constant
        return float(sympy_expr)
    else:
        raise TypeError(f"Unsupported SymPy expression type: {type(sympy_expr)}")


def symbolic_jacobian(model: Model) -> sp.Matrix:
    """
    Compute the Jacobian matrix of the ODE system in a model.

    Args:
        model: Model containing state variables and equations

    Returns:
        SymPy Matrix representing the Jacobian

    Raises:
        ValueError: If model has no state variables or equations
    """
    # Get all state variables
    state_vars = []
    for name, var in model.variables.items():
        if var.type == 'state':
            state_vars.append(name)

    if not state_vars:
        raise ValueError("Model has no state variables")

    if not model.equations:
        raise ValueError("Model has no equations")

    # Create symbol map for all variables
    symbol_map = {}
    for var_name in model.variables.keys():
        symbol_map[var_name] = sp.Symbol(var_name)

    # Extract right-hand sides of differential equations
    # Assume equations are of the form d(state_var)/dt = rhs
    rhs_expressions = []

    for equation in model.equations:
        # Convert equation to SymPy
        lhs_sympy = to_sympy(equation.lhs, symbol_map)
        rhs_sympy = to_sympy(equation.rhs, symbol_map)

        # Check if this is a differential equation
        if isinstance(lhs_sympy, sp.Derivative):
            # Extract the function being differentiated
            diff_var = lhs_sympy.args[0]
            if str(diff_var) in state_vars:
                rhs_expressions.append(rhs_sympy)
        else:
            # For non-differential equations, check if lhs is a state variable
            if str(lhs_sympy) in state_vars:
                # Treat as d(lhs)/dt = rhs
                rhs_expressions.append(rhs_sympy)

    if not rhs_expressions:
        # If no differential equations found, use all equation RHS
        for equation in model.equations:
            rhs_sympy = to_sympy(equation.rhs, symbol_map)
            rhs_expressions.append(rhs_sympy)

    # Ensure we have enough expressions
    while len(rhs_expressions) < len(state_vars):
        # Pad with zeros if needed
        rhs_expressions.append(sp.sympify(0))

    # Take only the first n expressions where n = number of state variables
    rhs_expressions = rhs_expressions[:len(state_vars)]

    # Compute Jacobian: J[i,j] = ∂(rhs_i)/∂(state_var_j)
    jacobian_elements = []
    for rhs in rhs_expressions:
        row = []
        for var_name in state_vars:
            var_symbol = symbol_map[var_name]
            partial_derivative = sp.diff(rhs, var_symbol)
            row.append(partial_derivative)
        jacobian_elements.append(row)

    return sp.Matrix(jacobian_elements)